tests/repositories/test_http_repository.py

```diff
@@ -153,3 +153,71 @@ def test_get_info_from_wheel_state_sequence(mocker: MockerFixture) -> None:
     repo._get_info_from_wheel(link)
     assert mock_metadata_from_wheel_url.call_count == 5
     assert mock_download.call_count == 4
+
+
+@pytest.mark.parametrize(
+    "mock_hashes",
+    [
+        None,
+        {"sha256": "e216b70f013c47b82a72540d34347632c5bfe59fd54f5fe5d51f6a68b19aaf84"},
+        {"md5": "be7589b4902793e66d7d979bd8581591"},
+    ],
+)
+def test_calculate_sha256(
+    mocker: MockerFixture, mock_hashes: dict[str, Any] | None
+) -> None:
+    filename = "poetry_core-1.5.0-py3-none-any.whl"
+    filepath = MockRepository.DIST_FIXTURES / filename
+    mock_download = mocker.patch(
+        "poetry.repositories.http_repository.download_file",
+        side_effect=lambda _, dest, *args, **kwargs: shutil.copy(filepath, dest),
+    )
+    domain = "foo.com"
+    link = Link(f"https://{domain}/{filename}", hashes=mock_hashes)
+    repo = MockRepository()
+
+    calculated_hash = repo.calculate_sha256(link)
+
+    assert mock_download.call_count == 1
+    assert (
+        calculated_hash
+        == "sha256:e216b70f013c47b82a72540d34347632c5bfe59fd54f5fe5d51f6a68b19aaf84"
+    )
+
+
+def test_calculate_sha256_defaults_to_sha256_on_md5_errors(
+    mocker: MockerFixture,
+) -> None:
+    suppressed_value_error = False
+
+    def mock_hashlib_md5_error() -> None:
+        nonlocal suppressed_value_error
+        suppressed_value_error = True
+        raise ValueError(
+            "[digital envelope routines: EVP_DigestInit_ex] disabled for FIPS"
+        )
+
+    filename = "poetry_core-1.5.0-py3-none-any.whl"
+    filepath = MockRepository.DIST_FIXTURES / filename
+    mock_download = mocker.patch(
+        "poetry.repositories.http_repository.download_file",
+        side_effect=lambda _, dest, *args, **kwargs: shutil.copy(filepath, dest),
+    )
+    mock_hashlib = mocker.patch("hashlib.md5", side_effect=mock_hashlib_md5_error)
+
+    domain = "foo.com"
+    link = Link(
+        f"https://{domain}/{filename}",
+        hashes={"md5": "be7589b4902793e66d7d979bd8581591"},
+    )
+    repo = MockRepository()
+
+    calculated_hash = repo.calculate_sha256(link)
+
+    assert suppressed_value_error
```

Comment:
Is this really necessary? Isn't it enough to check `mock_hashlib.call_count`? (This nonlocal thing feels a bit like a code smell.)

Reply:
Since the error is suppressed there's no way to know to know that the ValueError was raised (i.e., `with pytest.raises(ValueError)`).

It _should_ indeed be enough to assert that `mock_hashlib.call_count` was called though, with the assumption that the mock is working as expected - which is what I was avoiding by introducing a `nonlocal` variable (see [PEP 3104](https://peps.python.org/pep-3104/)) to test the error was raised but suppressed.

I can remove it though - just let me know if you still think it's unnecessary (:

Btw, just renamed this mocked function to `mock_hashlib_md5` to be more specific.

Reply:
After sleeping on it, I think I'm ok with it. Maybe, just rename `suppressed_value_error` to `raised_value_error` because that's what we actually care about?

Reply:
Done @radoering